---
title: "EdgeR_GLM_justheats_justcontrols"
author: "Melissa Naugle"
date: "5/20/2021"
output: html_document
---


#Used with Barshis transcriptome reference
#All 5 Sites

## First just controls

#Setup
Load packages and install BiocManager if needed

```{r setup}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("edgeR")
#install.packages("statmod")
library("edgeR")
library("statmod")
library("data.table")

setwd("~/Desktop/GitHub/RNAseq_allsites_Barshisreference/")
```

Import Data
Note: must open matrix and write 'gene' as first column header 

```{r import_data}
#Import gene count data and view the first few rows
#had to add 'gene' to first header row on original txt file to prevent weird error
countsTable <- read.table(file="../data/RSEM/RSEM.isoform.counts.matrix", row.names="gene", sep="\t", header=TRUE)
head(countsTable)

#just select controls
countsTable <- countsTable[,grepl("_E_", colnames(countsTable))]
head(countsTable)

```

Import grouping factors

Note: make sure reps txt file matches order of counts matrix

```{r groupingfactors}
#import grouping factor and create samples group matrix
colnames(countsTable)
reps <- read.table("coral_samples_allsites_controls.txt", header = F)
reps
```

Setup grouping factors and create List 


```{r create list}
group <- reps$V1

#create 'DGE list' object
#don't include group yet? #but r blog says to include so we include
#check to make sure row names match group names
y <- DGEList(counts=countsTable, group = group)

names(countsTable)
y

#MAKE SURE SAME ORDER ON REPS SHEET AS IN COUNTS MATRIX
colnames(y) <- reps$V2


```


Filtering and normalization

TMM normalization is performed to eliminate composition biases between libraries.

14148 after filtering
19348 filtered out

```{r norm_and_write_list}

#filter out lowly expressed genes
keep <- filterByExpr(y)

y <- y[keep,,keep.lib.sizes=FALSE]

#view summary of normalized counts
summary(keep)

#normalize library sizes (use trimmed mean m values to eliminate composition biases)
y <- calcNormFactors(y)

#not included in Jacoby's code 
#write normalized
normList <- cpm(y, normalized.lib.sizes = T)
nrow(normList)
write.table(normList, file = "../GLM_edgeR_outputs/glmQLF_normalizedCounts_controls.csv", sep=",", row.names=T)


```

MD plot to verify 

The performance of the TMM normalization procedure can be examined using meandifference (MD) plots. This visualizes the library size-adjusted log-fold change between two libraries (the difference) against the average log-expression across those libraries (the mean). The following MD plot is generated by comparing sample 1 against an artificial library constructed from the average of all other samples.

Ideally, the bulk of genes should be centred at a log-fold change of zero. This indicates that any composition bias between libraries has been successfully removed. This quality check should be repeated by constructing a MD plot for each sample.

```{r plots_to_verify}
#Verify TMM normalization using a MD plot
jpeg("../GLM_edgeR_outputs/glmQLF_plotNormalizedMD_controls.jpg")
plotMD(cpm(y, log=TRUE), column=1) 
abline(h=0, col="red", lty=2, lwd=2) 
dev.off()
```

Make MDS plot


default settings for MDS plot: 
points <- c(15,15,16,16,17,17)
colors <- rep(c("blue", "red"), 3)

```{r MDSplot}
#Use a MDS plot to visualizes the differences between samples
#shape is site and color is treatment
colors <- c("red","yellow","green","blue", "purple")
points <- rep(c(17, 16), 5)
#Write plot with legend to file
jpeg("../GLM_edgeR_outputs/controls/glmQLF_plotMDS_controls.jpg")
plotMDS(y, col=colors[group], pch=points[group], cex = 1.75, xlab = "PC1", ylab = "PC2") 
legend("topleft", legend=levels(group), pch=points, col=colors, ncol=2, cex = 0.75)
dev.off()


```


#-----------------------------------------------------------------------------------------

#Now just heats 

Import Data
Note: must open matrix and write 'gene' as first column header 

```{r import_data}
rm(list = ls())
#Import gene count data and view the first few rows
#had to add 'gene' to first header row on original txt file to prevent weird error
countsTable <- read.table(file="../data/RSEM/RSEM.isoform.counts.matrix", row.names="gene", sep="\t", header=TRUE)
head(countsTable)


#just select controls
countsTable <- countsTable[,grepl("_H_", colnames(countsTable))]
head(countsTable)
```

Import grouping factors

Note: make sure reps txt file matches order of counts matrix

```{r groupingfactors}
colnames(countsTable)
#import grouping factor and create samples group matrix
reps <- read.table("coral_samples_allsites_heat.txt", header = F)
reps
```

Setup grouping factors and create List 


```{r create list}
group <- reps$V1

#create 'DGE list' object
#don't include group yet? #but r blog says to include so we include
#check to make sure row names match group names
y <- DGEList(counts=countsTable, group = group)

names(countsTable)
y

#MAKE SURE SAME ORDER ON REPS SHEET AS IN COUNTS MATRIX
colnames(y) <- reps$V2


```


Filtering and normalization


15109 after filtering
18387 filtered out

```{r norm_and_write_list}

#filter out lowly expressed genes
keep <- filterByExpr(y)

y <- y[keep,,keep.lib.sizes=FALSE]

#view summary of normalized counts
summary(keep)

#normalize library sizes (use trimmed mean m values to eliminate composition biases)
y <- calcNormFactors(y)

#not included in Jacoby's code 
#write normalized
normList <- cpm(y, normalized.lib.sizes = T)
nrow(normList)
write.table(normList, file = "../GLM_edgeR_outputs/heats_only/glmQLF_normalizedCounts_heatsonly.csv", sep=",", row.names=T)


```

MD plot to verify 

The performance of the TMM normalization procedure can be examined using meandifference (MD) plots. This visualizes the library size-adjusted log-fold change between two libraries (the difference) against the average log-expression across those libraries (the mean). The following MD plot is generated by comparing sample 1 against an artificial library constructed from the average of all other samples.

Ideally, the bulk of genes should be centred at a log-fold change of zero. This indicates that any composition bias between libraries has been successfully removed. This quality check should be repeated by constructing a MD plot for each sample.

```{r plots_to_verify}
#Verify TMM normalization using a MD plot
jpeg("../GLM_edgeR_outputs/heats_only/glmQLF_plotNormalizedMD_heat.jpg")
plotMD(cpm(y, log=TRUE), column=1) 
abline(h=0, col="red", lty=2, lwd=2) 
dev.off()
```

Make MDS plot


default settings for MDS plot: 
points <- c(15,15,16,16,17,17)
colors <- rep(c("blue", "red"), 3)

```{r MDSplot}
#Use a MDS plot to visualizes the differences between samples
#shape is site and color is treatment
colors <- c("red","yellow","green","blue", "purple")
points <- rep(c(17, 16), 5)
#Write plot with legend to file
jpeg("../GLM_edgeR_outputs/heats_only/glmQLF_plotMDS_heatsonly.jpg")
plotMDS(y, col=colors[group], pch=points[group], cex = 1.75, xlab = "PC1", ylab = "PC2") 
legend("topleft", legend=levels(group), pch=points, col=colors, ncol=2, cex = 0.75)
dev.off()


```

