---
title: "Heatmap"
author: "Melissa Naugle"
date: "4/26/2021"
output: html_document
---

##Heats only

#1. Set WD and load packages

```{r setup}
setwd("/Users/Melissa/Desktop/GitHub/RNAseq_allsites_Barshisreference/data/Heats_only/")
library(tidyverse) 
library(reshape2)  
library(ggplot2)   
library(ggdendro)
library(gplots)
library(RColorBrewer)
```

#2. Read in data, check for duplicates, cluster, and reshape 

```{r data}
# load data
data <- read.table("diffExpr.P0.05_C2.matrix.log2.centered.dat", header=TRUE, sep = "\t")
head(data)
#how many contigs
nrow(data) 

#Check for row duplicates
data[duplicated(data) | duplicated(data, fromLast=TRUE), ]

#Re-organize data 
#cluster rows
row.order <- hclust(dist(data))$order 
#cluster columns by sample order 
names(data)
col.order <-c(1,2,3,4, #coco E
              23,24,25,26, #can E
              9,10,11,12, #alu E
              31,32,33,34, #vat E
              16,17,18, #tele E
              5,6,7,8, #coco H
              27,28,29,30, #can H
              13,14,15, #alu H
              35,36,37,38, #vat H
              19,20,21,22) #tele H
#re-order matrix according to clustering
#dat_clust <- data[row.order, col.order] # re-order matrix according to clustering
#reshape into data frame
#df_molten_dat <- melt(as.matrix(dat_clust)) 
#names(df_molten_dat)[c(1:2)] <- c("Trinity_ID", "treatment")
#df_molten_dat
```
##Heatmap with dendrogram 


```{r}

data_betterlabels <- data
names(data_betterlabels)
names(data_betterlabels) <- c("Coconut Point 1" , "Coconut Point 2" , "Coconut Point 3"  ,"Coconut Point 4"  ,"Faga'alu 1",  "Faga'alu 4"  ,"Faga'alu 6", "Faga'tele 1" ,"Faga'tele 3", "Faga'tele 5", "Faga'tele 6", "Cannery 2" ,  "Cannery 3",   "Cannery 5"  , "Cannery 6" ,  "Vatia 1",   "Vatia 3" ,  "Vatia 4"  , "Vatia 6")

my_palette <- colorRampPalette(c("purple", "black", "yellow"))

col_breaks = c(seq(-6,-2,length=100), # for purple
seq(-1.99,1.99,length=100),  # for black
seq(2,6,length=100)) #yellow


jpeg(file = "../../Figures/Classic_EdgeR/heats_only/heatmap_p0.05_heatsonly_07_18_21.jpg")

heatmap.2(as.matrix(data_betterlabels), dendrogram = c("column"), col=my_palette, trace="none", 
          labRow=FALSE, ylab="", xlab=" ", breaks = col_breaks, density.info="none", key.xlab="Log Fold Change",
          key.title =" ",margins =c(12,9))
dev.off()
```





#below is old heatmap - 

#3. Make heatmap

```{r heatmap}
#Find min and max of value column in data frame to determine range of heatmap

max <- max(df_molten_dat$value)
min <- min(df_molten_dat$value)
mean <- mean(df_molten_dat$value)
IQR(df_molten_dat$value)

#plot heatmap
g <- ggplot(df_molten_dat, aes(x=treatment,y=Trinity_ID)) +
  geom_tile(aes(fill = value)) +
  scale_fill_gradient2(
  #set colors
  low="blue", mid="black", high="orange", 
  #set midpoint, max, min, can change these to change the color scale
  midpoint=0, limits=c(-5,5)) + 
  ylab("Genes") +
  xlab("") +
  ggtitle("") +
  theme(legend.title = element_text(size = 10),
        legend.text = element_text(size = 12),
        plot.title = element_text(size=10),
        axis.title = element_text(size=14,face="bold"),
        #change x text size
        axis.text.x = element_text(angle = 90, hjust = 1),
        #make y text tiny. Can make bigger if you want to read contig names
        axis.text.y = element_text(angle = 20, size = 0))+
  labs(fill = "Log Fold Change")

g

```
