---
title: "WGCNA"
author: "Melissa Naugle"
date: "5/19/2021"
output: html_document
---


This is based on the WGCNA tutorial as created by The University of Texas at Austin found here: https://wikis.utexas.edu/display/bioiteam/Clustering+using+WGCNA

Files for this tutorial (in the WGCNA.zip folder) were dowloaded from here : https://wikis.utexas.edu/display/bioiteam/WGCNA+files

Also based on Holly's WGCNA tutorial: https://github.com/hdoerr/Rockfish_upwelling_DGE/blob/master/scripts/Analyses_and_Figures/WGCNA/WGCNA_Analysis.Rmd

#Before running the following tutorial, make sure that the following required packages are installed and that you have set the proper working directory.
```{r}
#Only run the following to install BiocManager if not already installed
#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
#BiocManager::install()
# Only run the following commands once to install WGCNA and flashClust on your computer
#BiocManager::install("WGCNA")
#install.packages("flashClust") 
# Load WGCNA and flashClust libraries every time you open R
library(WGCNA)
library(flashClust)
library("edgeR")
library("statmod")
library("data.table")
#Set your current working directory (where all your files are)
setwd("/Users/Melissa/Desktop/GitHub/RNAseq_allsites_Barshisreference/WGCNA/") 
```

# Step 1: Upload data into R and reformat for WGCNA
use: ../GLM_edgeR_outputs/glmQLF_normalizedCounts.csv

These counts have been normalized and filtered using filtbyexpr() in EdgeR in this script: https://github.com/melissanaugle/RNAseq_allsites_Barshisreference/blob/main/DGE_workflow/DGE_GLM_barshisreference.Rmd
18387 filtered out of 33496 total genes, leaving 15109 after filtering

Filtering deets: as described in (Chen et al. 2016)
-min.count = 10 (Minimum count required for at least some samples), 
-min.total.count = 15 (Minimum total count required)
-large.n = 10 (Number of samples per group that is considered to be “large”.)
-min.prop = 0.7 (Minimum proportion of samples in the smallest group that express the gene)

Normalization deets: trimmed mean m values to eliminate composition biases

This type of filtering and normalization means that the normalized counts are diff for the heats only v controls only v all samples. Is this okay? I suspect it is but they say defs DO NOT filter by DGE 

```{r}
# This creates an object called "datExpr" that contains the normalized counts file output from DESeq2
datExpr = read.csv("../GLM_edgeR_outputs/heats_only/glmQLF_normalizedCounts_heatsonly.csv")
# "head" the file to preview it
head(datExpr) # You see that genes are listed in a column named "X" and samples are in columns
```

Manipulate the file so it matches the format that WGCNA needs: 

commented out first two lines- this keeps contig IDs as column headers, I think this is better/okay?

```{r}
#row.names(datExpr) = datExpr$X
#datExpr$X = NULL
datExpr = as.data.frame(t(datExpr)) # now samples are rows and genes are columns
dim(datExpr) # 38 samples (or 19 for just heat/control) and 15109 genes (or fewer for heat/control)
head(datExpr)

row.names(datExpr)
```

We want to check if there are any gene outliers by running the following: 
```{r}
gsg = goodSamplesGenes(datExpr, verbose = 3)
gsg$allOK
```

## Do not run if 'TRUE'
If the last statement returns TRUE, all genes have passed the cuts. If not, we remove the offending genes and samples from the data with the following:

Passed, so do not run 
```{r}
if (!gsg$allOK)
   {if (sum(!gsg$goodGenes)>0)
       printFlush(paste("Removing genes:", paste(names(datExpr)[!gsg$goodGenes], collapse= ", ")));
       if (sum(!gsg$goodSamples)>0)
           printFlush(paste("Removing samples:", paste(rownames(datExpr)[!gsg$goodSamples], collapse=", ")))
       datExpr= datExpr[gsg$goodSamples, gsg$goodGenes]
       }
 
```


Create an object called "datTraits" that contains your trait data
```{r}
datTraits = read.csv("Heats_only_5202021/phenotypedata_justheats_05212021.csv")
head(datTraits)
nrow(datTraits)
#Then form a data frame analogous to expression data that will hold the clinical traits.
rownames(datTraits) = datTraits[,1]
datTraits$Sample = NULL
table(rownames(datTraits)==rownames(datExpr)) #should return TRUE if datasets align correctly, otherwise your names are out of order
head(datTraits)
```


You have finished uploading and formatting expression and trait data. Expression data is in datExpr, corresponding traits are datTraits. You can save these data into one file if you'd like, but you will not need to load it for the continuing code to run.

```{r}
#save(datExpr, datTraits, file="SamplesAndTraits.RData")
#load("SamplesAndTraits.RData")
```

At this point you will need to identify sample outliers and choose a soft threshold power. These are easy to do and are well documented in the online tutorials.  It is suggested that you cluster samples by expression to identify any outliers before this step. This is provided in the attached R script.

Tutorial note: From this plot, we would choose a power of 18 because it's the lowest power for which the scale free topology index reaches 0.90 (red line).

Melissa: scale free topology index doesn't pass ~0.82. Is that okay? Continue with 14 for now 

Also, sometimes weird stuff happens here and you get this error: "number of items to replace is not a multiple of replacement length", just restart R and all is okay 

https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html for soft threshold recs 

```{r}
# Choose a soft threshold power ------------------------------------
  
powers = c(c(1:10), seq(from =10, to=30, by=1)) #choosing a set of soft-thresholding powers
sft = pickSoftThreshold(datExpr, powerVector=powers, verbose =5, networkType="signed") #call network topology analysis function
  
sizeGrWindow(9,5)
par(mfrow= c(1,2))
cex1=0.9
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], xlab= "Soft Threshold (power)", ylab="Scale Free Topology Model Fit, signed R^2", type= "n", main= paste("Scale independence"))
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], labels=powers, cex=cex1, col="red") 
abline(h=0.9, col="red")
plot(sft$fitIndices[,1], sft$fitIndices[,5], xlab= "Soft Threshold (power)", ylab="Mean Connectivity", type="n", main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1, col="red")
 
```

#Step 2: Construct a gene co-expression network and identify modules

Holly chose 14, tutorial chose 18 


```{r}
#build a adjacency "correlation" matrix
enableWGCNAThreads()
softPower = 14
adjacency = adjacency(datExpr, power = softPower, type = "signed") #specify network type
head(adjacency)
```

Construct Networks- -----------------------------

```{r}
#translate the adjacency into topological overlap matrix and calculate the corresponding dissimilarity:
TOM = TOMsimilarity(adjacency, TOMType="signed") # specify network type
dissTOM = 1-TOM
```

Generate Modules --------------------------------------------------------

First, generate a clustered gene tree
```{r}
# Generate a clustered gene tree
geneTree = flashClust(as.dist(dissTOM), method="average")
plot(geneTree, xlab="", sub="", main= "Gene Clustering on TOM-based dissimilarity", labels= FALSE, hang=0.04)
#This sets the minimum number of genes to cluster into a module
minModuleSize = 30
dynamicMods = cutreeDynamic(dendro= geneTree, distM= dissTOM, deepSplit=2, pamRespectsDendro= FALSE, minClusterSize = minModuleSize)
dynamicColors= labels2colors(dynamicMods)
#Calculate eigengenes
MEList= moduleEigengenes(datExpr, colors= dynamicColors,softPower = softPower)
MEs= MEList$eigengenes
#Calculate dissimilarity of module eigengenes
MEDiss= 1-cor(MEs)
#Cluster module eigengenes
METree= flashClust(as.dist(MEDiss), method= "average")
#save(dynamicMods, MEList, MEs, MEDiss, METree, file= "Network_allSamples_signed_RLDfiltered.RData")
```

Then, this plots the tree showing how the eigengenes cluster together
```{r}
#plots tree showing how the eigengenes cluster together
#INCLUE THE NEXT LINE TO SAVE TO FILE
#pdf(file="clusterwithoutmodulecolors.pdf")
sizeGrWindow(7,6)
plot(METree, main= "Clustering of module eigengenes", xlab= "", sub= "")
#set a threhold for merging modules. In this example we chose a cut height of 0.25, corresponding to crrelation of 0.75, to merge
MEDissThres = 0.4
#Plot the cut line into the dendogram
abline(h=MEDissThres, col = "red")
#Call an automatic merging function
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose =3)
#Ther merged module colors
mergedColors = merge$colors
#Eigengenes of the new merged modules:
mergedMEs = merge$newMEs
#INCLUE THE NEXT LINE TO SAVE TO FILE
#dev.off()
```

And finally, this plots the dendogram with the module colors below it.
```{r}
#plot dendrogram with module colors below it
#INCLUE THE NEXT LINE TO SAVE TO FILE
pdf(file="cluster_controls.pdf")
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors), c("Dynamic Tree Cut", "Merged dynamic"), dendroLabels= FALSE, hang=0.03, addGuide= TRUE, guideHang=0.05)
moduleColors = mergedColors
colorOrder = c("grey", standardColors(50))
moduleLabels = match(moduleColors, colorOrder)-1
MEs = mergedMEs
#INCLUE THE NEXT LINE TO SAVE TO FILE
dev.off()
```

#Step 3: Relate modules to external traits.

Correlate traits --------------------------------------------------------

First, define the number of genes and samples:
```{r}
#Define number of genes and samples
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)
#Recalculate MEs with color labels
MEs0 = moduleEigengenes(datExpr, moduleColors)$eigengenes
MEs = orderMEs(MEs0)
moduleTraitCor = cor(MEs, datTraits, use= "p")
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)
```


Then, print the correlation heatmap between modules and traits

```{r}
#sizeGrWindow(15, 15)
#This will display correlations and their p-values
textMatrix= paste(signif(moduleTraitCor, 2), "\n(", 
                         signif(moduleTraitPvalue, 2), 
                         ")", sep= "");
#textMatrix = formatC(moduleTraitCor, digits = 2, format = "f")
dim(textMatrix)= dim(moduleTraitCor)
par(mar= c(6, 8.5, 3, 3));
```



Finally, display the correlation values with a heatmap plot:
```{r}
#display the correlation values with a heatmap plot
#sizeGrWindow(11,8)
#INCLUE THE NEXT LINE TO SAVE TO FILE
pdf(file="heatmap.pdf", height=11, width=8)
labeledHeatmap(Matrix= moduleTraitCor, 
            xLabels= names(datTraits), 
            yLabels= names(MEs), 
            ySymbols= names(MEs), 
            colorLabels= FALSE, 
            colors= blueWhiteRed(50), 
            textMatrix= textMatrix, 
            setStdMargins= TRUE, 
            cex.text= 0.75, 
            zlim= c(-1,1), 
            main= paste("Module-trait relationships"))
#INCLUE THE NEXT LINE TO SAVE TO FILE
dev.off()
```



And volia! You should have a beautiful clustered heatmap showing the correlation of traits and genes.

For further analysis, if you wanted to pull out genes belonging to a certain module, you can use the following command:
```{r}
names(datExpr)[moduleColors=="pink"]
```

#Further analysis and visual comparisons

1. Represent module trait correlations as a heatmap - version 2 = prettier 

```{r}
# module-trait correlations
quartz()
library(RColorBrewer)
modLabels=sub("ME","",names(MEs))
ps=signif(moduleTraitPvalue,1)
cors=signif(moduleTraitCor,2)
textMatrix = cors;
#paste(cors, "\n(",ps, ")", sep = "");
textMatrix[ps>0.05]="-"
dim(textMatrix) = dim(moduleTraitCor)
par(mar = c(6, 8.5, 3, 3));
```

```{r}
# Display the correlation values within a heatmap plot
pdf(file="Trait_heatmap_clean05202021.pdf", height=9, width=8)
labeledHeatmap(Matrix = moduleTraitCor,
xLabels = names(datTraits),
ySymbols = modLabels,
yLabels = modLabels,
colorLabels = FALSE,
colors = colorRampPalette(c("blue","lightblue","white","coral","red"))(50),
textMatrix = textMatrix,
setStdMargins = TRUE,
cex.text = 0.9,
zlim = c(-.8,.8),
main = paste("Module-Trait correlations05202021"))
dev.off()
```

2. Create a barplot to compare module sizes.

```{r}
# module size barplot
labelShift=300 # increase to move module size labels to the right
quartz()
par(mar = c(6, 8.5, 3, 3));
mct=table(moduleColors)
mct[modLabels]
x=barplot(mct[rev(modLabels)],horiz=T,las=1,xlim=c(0,8000),col=rev(modLabels))
text(mct[rev(modLabels)]+labelShift,y=x,mct[rev(modLabels)],cex=0.9) 
```

             
3. Calculate gene relationship to trait for important modules:

Need to change trait name in 3 spots below

```{r}
# Define variable weight containing the weight column of datTrait - leave weight as variable, but change names in first 2 commands
trait = "high_pollution_control"

weight = as.data.frame(datTraits$high_pollution_control); #change to your trait name of interest
names(weight) = "high_pollution_control"
# names (colors) of the modules
modNames = substring(names(MEs), 3)
geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");
geneTraitSignificance = as.data.frame(cor(datExpr, weight, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));
names(geneTraitSignificance) = paste("GS.", names(weight), sep="");
names(GSPvalue) = paste("p.GS.", names(weight), sep="");
```

Then plot the correlations for visual comparison

```{r}
#Gene-trait significance correlation plots
#windows()
moduleCols=c( "grey60") 
par(mfrow=c(1,4))
for (module in moduleCols) {
column = match(module, modNames);
moduleGenes = moduleColors==module;
#sizeGrWindow(7, 7);
par(mfrow = c(1,1));
verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
 abs(geneTraitSignificance[moduleGenes, 1]),
                 xlab = paste("ModMem in", module, "module"),
                 ylab = paste0("Gene Sig for ", trait),
                 main = paste("MM vs. GS\n"),
                 cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
   }     
```


4. Plotting massive table of all information - module membership, genes, gene names, etc.

Need to change module color under datOutput below

```{r}
annot=read.csv("../data/33496_MasterCombinedAnnotationTable.csv",header=TRUE,sep=",") #past_apr2014_iso2gene.tab; kb8_mar2014_iso2gene.tab
#match probes in the data set to the probe IDs in the annotation file
probes = names(datExpr)
probes2annot = match(probes,annot$Ourcontigname)
datGS.Traits=data.frame(cor(datExpr,datTraits,use="p"))
names(datGS.Traits)=paste("cor",names(datGS.Traits),sep=".")
datME=moduleEigengenes(datExpr,moduleColors)$eigengenes
datKME=signedKME(datExpr, datME, outputColumnName="MM.")
datOutput=data.frame(ProbeID=names(datExpr),annot[probes2annot,],moduleColors,datKME,datGS.Traits)
datOutput=datOutput[order((datOutput$MM.grey60),decreasing=T),]
write.table(datOutput,"AnnotatedNetworkAnalysisResults1kcontrols_test_05202021.csv",row.names=F,sep=",") 
```

5. Now you can save files containing the gene names and uniprot ID's for each interesting module.

```{r}
#Change the module color to create and save each file in the code below. 
#Get the corresponding uniprot id's
allUniprotIDs=annot$Uniprotmatch[probes2annot];
#Get the gene names
modGenes = names(datExpr)[moduleColors=="grey60"]
#Get their uniprot ID's
modUniprotIDs=allUniprotIDs[moduleColors=="grey60"];
#Write to a new file and save
fileName = data.frame(modGenes, modUniprotIDs)
write.table(fileName, "grey60_Module_genes_uniprot.csv", row.names=F, sep=",")
```


#--------------------------------------------------------------------------------------------------



#The following example illustrates 4 main ways of annotating columns and rows of a heatmap so you can customize your heatmap plot further.



- Copy and paste the whole example into an R session with an interactive plot window; alternatively, you may replace the command sizeGrWindow below by opening another graphical device such as pdf.

1. Generate a matrix to be plotted

```{r}
nCol = 8; nRow = 7;
mat = matrix(runif(nCol*nRow, min = -1, max = 1), nRow, nCol);
rowColors = standardColors(nRow);
colColors = standardColors(nRow + nCol)[(nRow+1):(nRow + nCol)];
rowColors;
colColors;
sizeGrWindow(9,7)
par(mfrow = c(2,2))
par(mar = c(4, 5, 4, 6));
```

2. Label rows and columns by text:

```{r}
labeledHeatmap(mat, xLabels = colColors, yLabels = rowColors, 
               colors = greenWhiteRed(50),
               setStdMargins = FALSE, 
               textMatrix = signif(mat, 2),
               main = "Text-labeled heatmap");
```

3. Label rows and columns by colors:

```{r}
rowLabels = paste("ME", rowColors, sep="");
colLabels = paste("ME", colColors, sep="");
labeledHeatmap(mat, xLabels = colLabels, yLabels = rowLabels,
               colorLabels = TRUE,
               colors = greenWhiteRed(50),
               setStdMargins = FALSE,
               textMatrix = signif(mat, 2),
               main = "Color-labeled heatmap");
```

4. Mix text and color labels:

```{r}
rowLabels[3] = "Row 3";
colLabels[1] = "Column 1";
labeledHeatmap(mat, xLabels = colLabels, yLabels = rowLabels,
               colorLabels = TRUE,
               colors = greenWhiteRed(50),
               setStdMargins = FALSE,
               textMatrix = signif(mat, 2), 
               main = "Mix-labeled heatmap");
```

5. Color labels and additional text labels

```{r}
rowLabels = paste("ME", rowColors, sep="");
colLabels = paste("ME", colColors, sep="");
extraRowLabels = paste("Row", c(1:nRow));
extraColLabels = paste("Column", c(1:nCol));
```

6. Extend margins to fit all labels

```{r}
par(mar = c(6, 6, 4, 6));
labeledHeatmap(mat, xLabels = colLabels, yLabels = rowLabels,
               xSymbols = extraColLabels,
               ySymbols = extraRowLabels,
               colorLabels = TRUE,
               colors = greenWhiteRed(50),
               setStdMargins = FALSE,
               textMatrix = signif(mat, 2),
               main = "Text- + color-labeled heatmap");
```
